system_tests/program_test.go

```diff
@@ -2618,3 +2623,115 @@ func TestRepopulateWasmLongTermCacheFromLru(t *testing.T) {
+	Require(t, err)
+	tx, err = arbOwner.SetL1PricingRewardRate(&auth, 0)
+	Require(t, err)
+	_, err = builder.L2.EnsureTxSucceeded(tx)
+	Require(t, err)
+	tx, err = arbOwner.SetL1PricePerUnit(&auth, big.NewInt(0))
+	Require(t, err)
+	_, err = builder.L2.EnsureTxSucceeded(tx)
+	Require(t, err)
+	tx, err = arbOwner.SetPerBatchGasCharge(&auth, 0)
+	Require(t, err)
+	_, err = builder.L2.EnsureTxSucceeded(tx)
+	Require(t, err)
+	tx, err = arbOwner.SetAmortizedCostCapBips(&auth, 0)
+	Require(t, err)
+	_, err = builder.L2.EnsureTxSucceeded(tx)
+	Require(t, err)
+
+	multiAddr := deployWasm(t, ctx, auth, builder.L2.Client, rustFile("multicall"))
+
+	argsMulticall := func(numberOfStores int) []byte {
+		args := multicallEmptyArgs()
+		for i := 0; i < numberOfStores; i++ {
+			key := testhelpers.RandomHash()
+			val := testhelpers.RandomHash()
+			args = multicallAppendStore(args, key, val, false, true)
+		}
+		return args
+	}
+
+	// Successful transaction to multicall
+	args := argsMulticall(50)
+	tx = builder.L2Info.PrepareTxTo("Owner", &multiAddr, uint64(2210000), nil, args)
+	err = builder.L2.Client.SendTransaction(ctx, tx)
+	Require(t, err)
+	receipt, err := builder.L2.EnsureTxSucceeded(tx)
+	Require(t, err)
+	if receipt.GasUsedForL1 != 0 {
+		t.Fatalf("expected 0 gas used, got %d", receipt.GasUsedForL1)
+	}
+	receipt, err = WaitForTx(ctx, testClient2ndNode.Client, tx.Hash(), time.Second*15)
+	Require(t, err)
+	if receipt.GasUsedForL1 != 0 {
+		t.Fatalf("expected 0 gas used, got %d", receipt.GasUsedForL1)
+	}
+
+	// Failed transaction to multicall.
+	// The transaction will fail during storage flush.
+	args = argsMulticall(200)
+	tx = builder.L2Info.PrepareTxTo("Owner", &multiAddr, uint64(2210000), nil, args)
+	err = builder.L2.Client.SendTransaction(ctx, tx)
+	Require(t, err)
+	receipt, err = builder.L2.EnsureTxSucceeded(tx)
+	if err == nil || err.Error() != fmt.Sprintf("out of gas for tx hash %v", tx.Hash().String()) {
+		t.Fatalf("expected out of gas error, got %v", err)
+	}
+	if receipt.GasUsedForL1 != 0 {
+		t.Fatalf("expected 0 gas used, got %d", receipt.GasUsedForL1)
+	}
+	receipt, err = WaitForTx(ctx, testClient2ndNode.Client, tx.Hash(), time.Second*15)
+	Require(t, err)
+	if receipt.GasUsedForL1 != 0 {
+		t.Fatalf("expected 0 gas used, got %d", receipt.GasUsedForL1)
+	}
+	blockNumberFailedTx := receipt.BlockNumber
+
+	wasmModuleRoot := currentRootModule(t)
+
+	// validation of previous block succeeds
+	_, _, err = builder.L2.ConsensusNode.StatelessBlockValidator.ValidateResult(
+		ctx,
+		arbutil.MessageIndex(blockNumberFailedTx.Uint64()-1),
+		false,
+		wasmModuleRoot,
+	)
+	Require(t, err)
+
+	// validation of block with failed transaction fails
+	_, _, err = builder.L2.ConsensusNode.StatelessBlockValidator.ValidateResult(
+		ctx,
+		arbutil.MessageIndex(blockNumberFailedTx.Uint64()),
+		false,
+		wasmModuleRoot,
+	)
+	if err == nil || !strings.Contains(err.Error(), "machine execution failed with error: missing requested preimage for hash") {
```

Comment:
Don't we want the second validation to also succeed?
I think we'd like to see the test failing without the fix here, and succeeding with the fix. Not sure what's best way to do it, maybe we can just see CI fail here and then merge in the fix (maybe even changing this PR target branch to [record-stylus-cache-flush](https://github.com/OffchainLabs/nitro/tree/record-stylus-cache-flush) and then seeing CI success?

Reply:
Agree. We want this test to fail right now, and merge this PR after we'vwe merged the solution

Reply:
Done.